Read runtime and devfile from files instead of workspace CR

amisevsk opened the pull request:
### What does this PR do?
Instead of reading the devfile and runtime annotation from the workspace custom resource (spec and status, respectively), expect files to be mounted to the container and read data from there.

File paths are configurable by env vars:
- `CHE_WORKSPACE_RUNTIME_JSON_PATH`
- `CHE_WORKSPACE_DEVFILE_YAML_PATH`

This means the che-rest-apis sidecar no longer needs rolebindings to run in a cluster.

An image from this changeset is pushed to `amisevsk/che-rest-apis:configmap`

Required for https://github.com/che-incubator/che-workspace-operator/issues/52

### Testing directly
These changes allow testing the image directly by building an image with test files included:

<details><summary>runtime.json</summary>
<p>

```json
{"activeEnv":"default","commands":[{"name":"mavenbuild","type":"exec","commandLine":"mvn clean install","attributes":{"actionReference":"","actionReferenceContent":"","componentAlias":"tools","machineName":"tools","workingDir":"${CHE_PROJECTS_ROOT}/web-java-spring-petclinic"}},{"name":"runwebapp","type":"exec","commandLine":"SPRING_DATASOURCE_URL=jdbc:mysql://db/petclinic\\\nSPRING_DATASOURCE_USERNAME=petclinic \\\nSPRING_DATASOURCE_PASSWORD=password\\\njava -jar -Dspring.profiles.active=mysql \\\n-Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=5005\\\ntarget/*.jar\n","attributes":{"actionReference":"","actionReferenceContent":"","componentAlias":"tools","machineName":"tools","workingDir":"${CHE_PROJECTS_ROOT}/web-java-spring-petclinic"}},{"name":"preparedatabase","type":"exec","commandLine":"/opt/rh/rh-mysql57/root/usr/bin/mysql-u root \u003c ${CHE_PROJECTS_ROOT}/web-java-spring-petclinic/src/main/resources/db/mysql/schema.sql\u0026\u0026\necho -e \"\\e[32mDone.\\e[0m Database petclinic was configured!\"\n","attributes":{"actionReference":"","actionReferenceContent":"","componentAlias":"mysql","machineName":"mysql","workingDir":""}}],"machines":{"che-machine-exec":{"attributes":{"plugin":"eclipse/che-machine-exec-plugin/latest","source":"tool"},"servers":{"che-machine-exec":{"attributes":{"cookiesAuthEnabled":"true","discoverable":"false","protocol":"ws","public":"true","secure":"true","type":"terminal"},"status":"RUNNING","url":"ws://workspaceeb5750a855d74f09-che-machine-exec-4444.apps-crc.testing"}}},"che-rest-apis":{"attributes":{"source":"tool"}},"mysql":{"attributes":{"source":"recipe"}},"theia-ide":{"attributes":{"plugin":"eclipse/che-theia/latest","source":"tool"},"servers":{"theia-dev":{"attributes":{"discoverable":"false","protocol":"http","public":"true","type":"ide-dev"},"status":"RUNNING","url":"http://workspaceeb5750a855d74f09-theia-dev-3130.apps-crc.testing"},"theia-proxy":{"attributes":{"cookiesAuthEnabled":"true","discoverable":"false","protocol":"http","public":"true","secure":"true","type":"ide"},"status":"RUNNING","url":"https://workspaceeb5750a855d74f09-theia-proxy-4400.apps-crc.testing"},"theia-redirect-1":{"attributes":{"discoverable":"false","protocol":"http","public":"true"},"status":"RUNNING","url":"http://workspaceeb5750a855d74f09-theia-redirect-1-13131.apps-crc.testing"},"theia-redirect-2":{"attributes":{"discoverable":"false","protocol":"http","public":"true"},"status":"RUNNING","url":"http://workspaceeb5750a855d74f09-theia-redirect-2-13132.apps-crc.testing"},"theia-redirect-3":{"attributes":{"discoverable":"false","protocol":"http","public":"true"},"status":"RUNNING","url":"http://workspaceeb5750a855d74f09-theia-redirect-3-13133.apps-crc.testing"}}},"tools":{"attributes":{"source":"recipe"}},"vscode-java":{"attributes":{"plugin":"redhat/java/latest","source":"tool"}}}}
```

</p>
</details>

<details><summary>devfile.yaml</summary>
<p>

```yaml
apiVersion: 1.0.0
projects:
  - name: web-java-spring-petclinic
    source:
      type: git
      location: "https://github.com/spring-projects/spring-petclinic.git"
components:
  - type: chePlugin
    id: redhat/java/latest
    memoryLimit: 1280MiB
  - type: dockerimage
    alias: tools
    image: quay.io/eclipse/che-java8-maven:nightly
    env:
      - name: JAVA_OPTS
        value: "-XX:MaxRAMPercentage=50.0 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10
          -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90
          -Dsun.zip.disableMemoryMapping=true -Xms20m -Djava.security.egd=file:/dev/./urandom
          -Duser.home=/home/user"
      - name: MAVEN_OPTS
        value: $(JAVA_OPTS)
    memoryLimit: 700Mi
    endpoints:
      - name: '8080/tcp'
        port: 8080
    mountSources: true
    volumes:
      - name: m2
        containerPath: /home/user/.m2
  - type: dockerimage
    alias: mysql
    image: centos/mysql-57-centos7
    env:
      - name: MYSQL_USER
        value: petclinic
      - name: MYSQL_PASSWORD
        value: password
      - name: MYSQL_DATABASE
        value: petclinic
      - name: PS1
        value: $(echo ${0})\\$
    memoryLimit: 300Mi
    endpoints:
      - name: 'db'
        port: 3306
        attributes:
          discoverable: "true"
          public: "false"
    mountSources: true
  - alias: theia-ide
    type: cheEditor
    id: eclipse/che-theia/latest
  - type: chePlugin
    id: eclipse/che-machine-exec-plugin/latest
commands:
  - name: maven build
    actions:
      - type: exec
        component: tools
        command: "mvn clean install"
        workdir: "${CHE_PROJECTS_ROOT}/web-java-spring-petclinic"
  - name: run webapp
    actions:
      - type: exec
        component: tools
        command: |
          SPRING_DATASOURCE_URL=jdbc:mysql://db/petclinic \
          SPRING_DATASOURCE_USERNAME=petclinic \
          SPRING_DATASOURCE_PASSWORD=password \
          java -jar -Dspring.profiles.active=mysql \
          -Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=5005 \
          target/*.jar
        workdir: ${CHE_PROJECTS_ROOT}/web-java-spring-petclinic
  - name: prepare database
    actions:
    - type: exec
      component: mysql
      command: |
        /opt/rh/rh-mysql57/root/usr/bin/mysql -u root < ${CHE_PROJECTS_ROOT}/web-java-spring-petclinic/src/main/resources/db/mysql/schema.sql &&
        echo -e "\e[32mDone.\e[0m Database petclinic was configured!"
  - name: Debug remote java application
    actions:
      - type: vscode-launch
        referenceContent: |
          {
          "version": "0.2.0",
          "configurations": [
            {
              "type": "java",
              "name": "Debug (Attach) - Remote",
              "request": "attach",
              "hostName": "localhost",
              "port": 5005
            }]
          }
```

</p>
</details>

```dockerfile
FROM amisevsk/che-rest-apis:configmap

COPY ./runtime.json /workspace/runtime.json
COPY ./devfile.yaml /workspace/devfile.yaml
```

```bash
# build image
docker build -t che-rest-apis:test .
docker run --rm -p 9999:9999 che-rest-apis:test
curl localhost:9999/api/workspace/id | jq .
```

sleshchenko commented on src/main/java/org/eclipse/che/incubator/crd/cherestapis/ApiService.java:
```diff
@@ -121,47 +132,20 @@ public WorkspaceDto getWorkspace(String workspaceId) {
+            runtimeObj = parseRuntime(runtimeJson);
+            LOGGER.debug(runtimeObj.toString());
+        } catch (JsonMappingException  e) {
+            throw new RuntimeException("Failed to read json TODO " + e.getMessage());
```
Please fix TODO before merging

amisevsk replied:
Woops :D

amisevsk committed:
Read runtime and devfile from files instead of workspace CR

Add properties

  che.workspace.runtime.json.path
  che.workspace.devfile.path

that define paths to files containing the runtime annotation and
devfile, respectively. Workspace DTO object is constructed from these
files rather than by reading the workspace's custom resource.

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

amisevsk commented:
Pushed changes to `amisevsk/che-rest-apis:v0.0.2` for now, created issue https://github.com/eclipse/che/issues/16636 for setting up CI/release process.

`amisevsk/che-rest-apis:latest` is the master branch, without this PR. I've also pushed it to `amisevsk/che-rest-apis:v0.0.1` in case we need to revert for whatever reason.
